### PR TITLE
Implemented required includes -- again

### PR DESCRIPTION
**Pull request by Johnlon**
Redone from scratch.
This includes additional tests proposed last time I raised a pull request.

### CODE REVIEW

**Comment by Johnlon** [comment]
Regards issue https://github.com/typesafehub/config/issues/121

**Comment by havocp** [comment]
This looks good to me, thank you - any comments from @2m / @patriknw / others?

File: HOCON.md [hunk 1]
```diff
@@ -1031,12 +1032,24 @@ get a value from a system property or from the reference
...
+If however an included resource is mandatory then the name of the
+included resource may be wrapped with `required()`, in which case
+file parsing will fail with an error if the resource cannot be resolved.
+
+The syntax for this is
+
+    include required("foo.conf")
+    include required(file("foo.conf"))
+    include required(classpath("foo.conf"))
+    include required(url("http://localhost/foo.conf"))
```
Comment (patriknw) [inline HOCON.md]: nice

**Comment by patriknw** [comment]
LGTM

**Comment by Johnlon** [comment]
2 years in the making

**Comment by havocp** [comment]
Thanks! Sorry about the dead ends. We got there in the end.

**Pull request merged**